core/src/mast/merger.rs

```diff
@@ -0,0 +1,655 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+
+use miden_crypto::hash::blake::Blake3Digest;
+
+use crate::mast::{
+    DecoratorId, EqHash, MastForest, MastForestDfsIter, MastForestError, MastNode, MastNodeId,
+};
+
+/// A type that allows merging [`MastForest`]s.
+///
+/// Merging two forests means combining all their constituent parts, i.e. [`MastNode`]s,
+/// [`Decorator`](crate::mast::Decorator)s and roots. During this process, any duplicates are
+/// removed. Additionally, [`MastNodeId`]s of nodes as well as [`DecoratorId`]s of decorators change
+/// and they are remapped to their new location.
+///
+/// For example, consider this representation of a forest's nodes:
+///
+/// ```text
+/// [Block(foo), Block(bar)]
+/// ```
+///
+/// If we merge another forest into it:
+///
+/// ```text
+/// [Block(bar), Call(0)]
+/// ```
+///
+/// then we would expect this forest:
+///
+/// ```text
+/// [Block(foo), Block(bar), Call(1)]
+/// ```
+///
+/// - The `Call` to the `bar` block was remapped to its new index (now 1, previously 0).
+/// - The `Block(bar)` was deduplicated any only exists once in the merged forest.
+///
+/// Note that there are convenience methods for merging on [`MastForest`] itself:
+/// - [`MastForest::merge`]
+/// - [`MastForest::merge_multiple`]
+pub(crate) struct MastForestMerger<'forest> {
+    mast_forest: &'forest mut MastForest,
+    // Internal indices needed for efficient duplicate checking.
+    node_id_by_hash: BTreeMap<EqHash, MastNodeId>,
+    hash_by_node_id: BTreeMap<MastNodeId, EqHash>,
+    decorators_by_hash: BTreeMap<Blake3Digest<32>, DecoratorId>,
+}
+
+impl<'forest> MastForestMerger<'forest> {
+    /// Creates a new merger from the given mutable reference to a [`MastForest`].
+    ///
+    /// This forest will be used as the base of the merger, meaning any calls to [`Self::merge`]
+    /// will mutate this `forest` in-place.
+    pub(crate) fn new(mast_forest: &'forest mut MastForest) -> Self {
+        let mut forest = Self {
+            node_id_by_hash: BTreeMap::new(),
+            hash_by_node_id: BTreeMap::new(),
+            decorators_by_hash: BTreeMap::new(),
+            mast_forest,
+        };
+
+        forest.build_index();
+
+        forest
+    }
+
+    /// Merges `other_forest` into the forest contained in self.
+    ///
+    /// This will mutate the contained forest in-place.
+    pub(crate) fn merge(&mut self, other_forest: &MastForest) -> Result<(), MastForestError> {
```

Comment:
I forget exactly where we intend to use this, but I believe we should also be returning the ID remappings (similar to how we do it in `MastForest::remove_nodes()`)? It basically depends on whether we want to allow users to interact with a `MastForest` (and save some `MastNodeId`s somewhere), and then merge that same forest into another one (effectively invalidating all previously retrieved IDs). *Ideally* though we don't allow this as managing id remappings can become quite messy, but I'm not sure if this is realistic.

cc @bobbinth @bitwalker

Reply:
In the context I am thinking about, we probably don't need to have remappings of all nodes, but we may need remappings of procedure roots. The scenario is basically where we have 2 or more `Library` structs and then we merge their `MastForest`s, but then we'd want to be able to figure out how procedure names map to procedure roots. But if the incremental effort of returning remappings for all node is not great, that would probably be better.

I think the main challenge here is that we may need 2 separate maps: one for how node IDs in the `self` forest changed, and the other one in how the IDs in the `other_forest` changed. But maybe there is a better way to handle it.

Reply:
One potential option is use tuple `(usize, MastNodeId)` as the key in the returned map, where the `usize` is the index of the forest (`0` for self, `1` for `other_forest`). So, the map would look something like:
```rust
BTreeMap<(usize, MastNodeId), MastNodeId>
```
Maybe forest indexes can be extended beyond 1 in case of `merge_multiple()`.

Reply:
Does this mean we should also expose a `MastForest::remap(&mut self, remappings: BTreeMap<MastNodeId, MastNodeId>` function that makes it easy to remap other MastForests with a given mapping? At least I think this functionality is not yet exposed anywhere.

I deliberately omitted the `usize` because the remap function wouldn't know which index to use, unless we pass it along. The point is that this couples the mapping function to exactly this type, and if we want to expose remappings from other sources (like the builder for some reason) then we would have to return `BTreeMap<(usize, MastNodeId), MastNodeId>` there as well to be consistent in our API. And for something like `remove_nodes` it might be odd to return such a map even if it operates on just a single tree.

So perhaps it is cleaner but less efficient if we return one `BTreeMap<MastNodeId, MastNodeId>` per forest (in the case of `merge_multiple`).

And this would also allow us to encapsulate this mapping into some more "opaque" type (like `MastForestIdMapping` or something which doesn't really expose any functionality) since I imagine users do not need to be able to create it themselves? It's only a container to pass the mapping from one source to the remap function. This would give us some leeway if we want to change the internal representation of the mapping it in the future (after we've committed to a stable API).

Not sure if this makes sense in the larger picture though.

Reply:
> And this would also allow us to encapsulate this mapping into some more "opaque" type (like `MastForestIdMapping` or something which doesn't really expose any functionality) since I imagine users do not need to be able to create it themselves? It's only a container to pass the mapping from one source to the remap function. This would give us some leeway if we want to change the internal representation of the mapping it in the future (after we've committed to a stable API).

I think it is a good idea to create a dedicated struct for remappings.

> I deliberately omitted the `usize` because the remap function wouldn't know which index to use, unless we pass it along. The point is that this couples the mapping function to exactly this type, and if we want to expose remappings from other sources (like the builder for some reason) then we would have to return `BTreeMap<(usize, MastNodeId), MastNodeId>` there as well to be consistent in our API. And for something like `remove_nodes` it might be odd to return such a map even if it operates on just a single tree.

Agreed - let's keep the return types consistent.

> So perhaps it is cleaner but less efficient if we return one `BTreeMap<MastNodeId, MastNodeId>` per forest (in the case of `merge_multiple`).

I'm fine with this - but in the case of a single `merge()`, we'd also need to return two maps, right?

> Does this mean we should also expose a `MastForest::remap(&mut self, remappings: BTreeMap<MastNodeId, MastNodeId>` function that makes it easy to remap other MastForests with a given mapping? At least I think this functionality is not yet exposed anywhere.

I'm note sure that's needed this function to be public (or at least i'm not seeing an external use case for it).

Reply:
If we encapsulate the mappings into a type without real functionality and we don't add an API which uses that type, then I don't think I quite understand what the use of this type would be. Would we add something later?

Reply:
With this type, we should be able to map node ID from a given old forest to a node ID in the merged forest. I don't think we need more functionality than this. We could use a `BTreeMap` for this, but as you've suggested, it may be good to not expose the underlying type in case we change it to something else later.

One potential use case for this is trying to merge two `Library` structs. These have maps from procedure names to their corresponding roots and so if we wanted to create a single library by merging two libraries, we'd need to be able to make the type of lookup I described above.